cooked-validators/src/Cooked/Attack.hs

```diff
@@ -0,0 +1,158 @@
+addLabel :: LabelConstrs x => x -> TxSkel -> TxSkel
+addLabel newlabel =
+  over
+    txLabelL
+    ( \case
+        TxLabel Nothing -> TxLabel $ Just newlabel
+        TxLabel (Just oldlabel) -> TxLabel $ Just (newlabel, oldlabel)
+    )
+
+-- | try to use the given function to modify the elements of the given list. If
+-- at least one modification is successful, return 'Just' the modified list,
+-- otherwise fail returning 'Nothing'.
+someJust :: (a -> Maybe a) -> [a] -> Maybe [a]
+someJust _ [] = Nothing
+someJust f (x : xs) = case f x of
+  Just y -> Just $ y : map (\a -> fromMaybe a (f a)) xs
+  Nothing -> (x :) <$> someJust f xs
+
+-- * Token duplication attack
+
+-- | A token duplication attack which modifies every 'Mints'-constraint of a
+-- 'TxSkel' that satisfies some conditions. This adds a 'DupTokenLbl' to the
+-- labels of the transaction using 'addLabel'.
+dupTokenAttack ::
+  -- | A function describing how the amount of minted tokens of an asset class
+  -- should be changed by the attack. The given function @f@ should probably satisfy
+  -- > f ac i == Just j -> i < j
+  -- for all @ac@ and @i@, i.e. it should increase in the minted amount.
+  -- If it does *not* increase the minted amount, or if @f ac i == Nothing@, the
+  -- minted amount will be left unchanged.
+  (L.AssetClass -> Integer -> Maybe Integer) ->
+  -- | The wallet of the attacker. Any extra tokens that were minted by the
+  -- transaction are paid to this wallet.
+  Wallet ->
+  Attack
+dupTokenAttack change attacker skel =
+  addLabel DupTokenLbl . paySurplusTo attacker
+    <$> mkAttack (mintsConstraintsT % valueL) increaseValue skel
+  where
+    increaseValue :: L.Value -> Maybe L.Value
+    increaseValue v =
+      case someJust
+        ( \(ac, i) -> case change ac i of
+            Nothing -> Nothing
+            Just j -> if i < j then Just (ac, j) else Nothing
+        )
+        (view flattenValueI v) of
+        Just l -> Just $ review flattenValueI l
+        Nothing -> Nothing
+
+    paySurplusTo :: Wallet -> TxSkel -> TxSkel
+    paySurplusTo w s = over outConstraintsL (paysPK (walletPKHash w) surplus :) s
+      where
+        surplus = txSkelInValue s <> Pl.negate (nonAdaValue $ txSkelOutValue s)
+
+data DupTokenLbl = DupTokenLbl
+  deriving (Eq, Show)
+
+-- * Datum hijacking attack
+
+-- | A datum hijacking attack, simplified: This attack tries to substitute a
+-- different recipient on 'PaysScript' constraints, but leaves the datum as it
+-- is. That is, it tests for careless uses of something like 'txInfoOutputs' in
+-- places where something like 'getContinuingOutputs' should be used. If this
+-- attack goes through, however, a "proper" datum hijacking attack that modifies
+-- the datum in a way that the (relevant part of) the
+-- 'toBuiltinData'-translation stays the same will also work. A
+-- 'DatumHijackingLbl' is added to the labels of the 'TxSkel' using 'addLabel'.
+datumHijackingAttack ::
```

Comment:
I wonder a few things here:

1. If we pass two predicates that have to be valid, why don't we pass a single predicate to the function?
    ```
    (TypedValidator a -> DatumType a -> Value -> Bool)
    ```

2. Should we additionally pass an integer to that function indicating which output it is? Or, alternatively, maybe the semantics should apply the datum hijacking to /the first/ output satisfying the selection predicate. I ask because we might want to apply the hijacking to a single output even if the transaction creates two identical outputs.

Reply:
I think I remember you or someone saying that some contracts rely on the order of outputs. So, I'm in favour of a mechanism to select the n-th matching output.